samples/performance/constant_data/constant_data_tutorial.md

```diff
@@ -0,0 +1,405 @@
+2. Allocating a **descriptor set** with a binding containing the `VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER_DYNAMIC` flag that then points to the **UBO** you just created.
+3. In our draw call, for each mesh, *dynamically* offset into the giant **uniform buffer object**.
+
+### **Performance**
+
+![dynamic ubo](images/dynamic_descriptor_set_performance.jpg)
+
+In the screenshot above (taken on an S10 with a Mali G76 GPU) we can see the load/store cycles stay roughly the same compared to [static uniform buffer objects](#uniform-buffer-objects). However, the frame time goes up from 16.7 ms to 20.9 ms. This is due to the extra time you need to spend every frame determining the dynamic offsets, that you need to send in the bind call (`vkCmdBindDescriptorSets`).
+
+## **Update-after-bind Descriptor Sets**
+
+### **Introduction**
+
+Traditionally, **descriptor sets** require updating before they are bound to a command buffer - any further updates after it is bound will invalidate the command buffer it is bound to. However, this can be considered an "overly cautious" restriction when we realise that the command buffer isn't actually executed until it's submitted on a queue. This is where newer versions of Vulkan have introduced the concept of "update after bind". 
+
+Essentially it adds in a binding flag to descriptor set layouts which allows the contents of the **descriptor set** to be updated up until the command buffer *is submitted to the queue*, rather than when the descriptor set is *bound to the command buffer*.
+
+***Note:** Update-after-bind bindings cannot be used with dynamic descriptor sets.*
+
+### **Performance**
+
+This should come with zero performance costs, and as a result this method is designed purely for offering flexibility to your codebase. 
+
+## **Buffer Object Arrays**
+
+### **Introduction**
+
+Another approach, which can be likened to a dynamic descriptor set, is a buffer object array. This is the concept of allocating all of your constant data *upfront* in a large buffer, and writing the entire buffer to a descriptor set. This means in any one shader invocation we have access to all of the model data for the entire scene, at the benefit of only needing to bind one descriptor set per entire draw call.
+
+You can use either a `uniform` or a `buffer` storage type in your shader code to achieve this. However, since `buffer`s can generally hold bigger amounts of data, this tutorial will use them. 
+
+***Note:** If deciding to use a `uniform`, then the size of the array needs to be defined at compile time. This can be achieved with a shader variant definition.*
+
+Here is an example of using a `buffer` in shader code:
+
+```
+layout(set = 0, binding = 1) buffer MeshArray
+{
+	mat4 model_matrices[];
+} mesh_array;
+```
+
+***Note:** If you plan to write to `buffer`s inside shaders, you have to query the physical device features and check that `vertexPipelineStoresAndAtomics` is enabled.*
```

Comment:
If you plan to write to buffers inside fragment shaders, you have to query the physical device features and check that fragmentStoresAndAtomics is enabled (vertexPipelineStoresAndAtomics when storing in vertex, tessellation, and geometry shaders).

Reply:
I've removed it altogether as writing is really relevant to constant data